qt: add option to disable controller applet

EBADBEEF opened the pull request:
- add checkbox to disable the controller applet UI
- when controller applet is disabled, use the yuzu-cmd fallback controller applet that applies controller config based on rules
- See https://github.com/yuzu-emu/yuzu/issues/8552 for some discussion

german77 commented:
You need to do a bit more than this. You need to satisfy the rules of the applet as a fallback. Otherwise you will stay in a infinite loop of the game asking for the applet

EBADBEEF commented:
@german77 Can you give me an example? Would yuzu-cmd get stuck in a loop in the same case too? If my reading of the source is correct, it uses the existing fallback handler used by yuzu-cmd if no applet is supplied in SetAppletFrontendSet - https://github.com/yuzu-emu/yuzu/blob/f99f618d45ad862c4bc23fc28c91d1c48218a3cb/src/core/hle/service/am/applets/applets.cpp#L250

german77 commented:
This is the set of rules the applet requires https://github.com/yuzu-emu/yuzu/blob/08d4e7c7aff6b4b798b9095c5715368918cdba0b/src/core/frontend/applets/controller.h#L20

Actually yes the cmd version has the same issue. You don't need to go that far mk8 has this issue. Just set `ui->buttonBox->setEnabled` to be always true so you can click the button and try to start a race with the incorrect number of players. To solve it you need to automatically connect controllers and set the correct type to satisfy with the rules. This can be done by accessing the EmulatedController directly

Brief example:
```
    npad_device = system.HIDCore().GetEmulatedController(npad_id);
    npad_device->Connect();
    npad_device->Disconnect();
    npad_device->SetNpadStyleIndex(Core::HID::NpadStyleIndex::ProController);
```

![applet](https://user-images.githubusercontent.com/5944268/214100061-fcfd68e0-fade-4be1-b2aa-87b58a0749e9.gif)

EBADBEEF commented:
I'm trying to understand what needs to be done. Is what you described different from what DefaultControllerApplet::ReconfigureControllers does? It walks through all the controllers available and makes them conform to the requested parameters using SetNpadStyleIndex (https://github.com/yuzu-emu/yuzu/blob/08d4e7c7aff6b4b798b9095c5715368918cdba0b/src/core/frontend/applets/controller.cpp#L19).

What else is missing? If we don't have enough controllers, what happens? Oh, `hid_core.available_controllers` is always 10. So I guess walks through all possible controllers and connects them (even if they have no real controller backing them).

I've only tested Mario Odyssey (2 players) and it doesn't seem to keep calling the controller applet. I see the following my logs:
```
[ 143.874133] Service.HID <Info> core/hle/service/am/applets/applet_controller.cpp:Initialize:59: Initializing Controller Applet.
[ 143.874143] Service.HID <Info> core/frontend/applets/controller.cpp:ReconfigureControllers:21: called, deducing the best configuration based on the given parameters!
```

If I start Mario Odyssey with one controller plugged in, switch to two player mode, connect second controller, I can successfully use 2-player mode. I don't see ReconfigureControllers being called when the new controller is plugged in.

FearlessTobi commented:
A solution I'd honestly like much more than this would be to actually implement a controller applet that can be navigated without a mouse, Or trying to at least make the QT applet useable with a controller.

Hixbot2 commented:
> A solution I'd honestly like much more than this would be to actually implement a controller applet that can be navigated without a mouse, Or trying to at least make the QT applet useable with a controller.

Interesting but at least in my case the issue with the applet is not just that it must be used with a mouse.   It is also that the applet is a separate window that requires focus when using, and then the main window need to be re-focused after the applet closes.  This doesn't play nice with some configurations, for example when using Moonlight on Android as the client device and Yuzu on Windows as the stream host.

Ebadbeef's idea to disable or auto-fill and close the applet according to pre-configured rules makes alot of sense in my view.

EBADBEEF commented:
@german77 I would appreciate another look at this. See earlier comment https://github.com/yuzu-emu/yuzu/pull/9663#issuecomment-1401233317 (quick summary: I believe the existing default controller applet already addresses all of your concerns).

I have since tested more games (including MK8) and haven't noticed any problems.

FearlessTobi commented:
> It is also that the applet is a separate window that requires focus when using, and then the main window need to be re-focused after the applet closes. 

To be honest I think this could just be fixed by returning the focus back to the main yuzu window after the applet has closed...

coolllman commented:
When will this PR be merged?

FearlessTobi commented:
It will be merged when the feedback has been addressed by the author.

EBADBEEF commented:
@FearlessTobi Do you mean your comments about wanting to navigate the applet with controller and/or handle focus better? I agree that would be good, but I would prefer to keep the option to not show a dialog. Can you convince me that the applet is actually useful?

My use case is a system with 1-4 bluetooth controllers (pro controller style) and displaying on a TV. I launch yuzu from a frontend purely with a controller. To quit yuzu I also use the controller (with a separate program listening for a specific button combo). I even have it set up to turn on and off the TV with the controller (haha). The controllers can get connected/disconnected at will. The keyboard and mouse is a crappy little wireless one that I keep just for emergencies. Maybe I should just be using yuzu-cmd? But then when I do need the UI it is annoying to switch back. The separate configs are also hard to manage. I can't remember but I might have had some other issues with yuzu-cmd.

Side note - sometimes a controller gets "disconnected" in yuzu and I have to go reconnect it in the UI. Mainly noticed on SSBU. I might try to look at that issue later (seems to happen with and without fancy controller applet).
